opt_einsum/paths.py

```diff
@@ -506,3 +596,257 @@ def greedy(inputs, output, size_dict, memory_limit=None):
+    """
+
+    def __init__(self, max_repeats=32, max_time=None, minimize='flops', cost_fn=None,
+                 temperature=1.0, rel_temperature=True, nbranch=8, executor=None):
+
+        if minimize not in ('flops', 'size'):
+            raise ValueError("`minimize` should be one of {'flops', 'size'}.")
+
+        self.max_repeats = max_repeats
+        self.max_time = max_time
+        self.minimize = minimize
+        self.cost_fn = cost_fn
+        self.better = _BETTER_FNS[minimize]
+        self.temperature = temperature
+        self.rel_temperature = rel_temperature
+        self.nbranch = nbranch
+        self.executor = executor
+
+        self.costs = []
+        self.sizes = []
+        self.best = {'cost': float('inf'), 'size': float('inf')}
+
+        # this keeps track of how many trials we have submitted so that (a)
+        # each trial has a different seed, and (b) the very first trial can be
+        # made to always be the standard greedy one
+        self._r0 = 0
+
+    @property
+    def choose_fn(self):
+        """The function that chooses which contraction to take - make this a
+        property so that ``temperature`` and ``nbranch`` etc. can be updated
+        between runs.
+        """
+        if self.nbranch == 1:
+            return None
+
+        return functools.partial(thermal_chooser, temperature=self.temperature,
+                                 nbranch=self.nbranch, rel_temperature=self.rel_temperature)
+
+    @property
+    def path(self):
+        """The best path found so far.
+        """
+        return ssa_to_linear(self.best['ssa_path'])
+
+    def __call__(self, inputs, output, size_dict, memory_limit):
+        import time
+
+        # start a timer?
+        if self.max_time is not None:
+            t0 = time.time()
+
+        args = (inputs, output, size_dict, self.choose_fn, self.cost_fn)
+        repeats = range(self._r0, self._r0 + self.max_repeats)
+        self._r0 += self.max_repeats
+
+        # create the trials lazily
+        if self.executor is not None:
+            # eagerly submit
+            fs = [self.executor.submit(_trial_ssa_path_and_cost, r, *args) for r in repeats]
+            # lazily retrieve
+            trials = (f.result() for f in fs)
+        else:
+            trials = (_trial_ssa_path_and_cost(r, *args) for r in repeats)
+
+        # assess the trials
+        for ssa_path, cost, size in trials:
+
+            # keep track of all costs and sizes
+            self.costs.append(cost)
+            self.sizes.append(size)
+
+            # check if we have found a new best
+            found_new_best = self.better(cost, size, self.best['cost'], self.best['size'])
+
+            if found_new_best:
+                self.best['cost'] = cost
+                self.best['size'] = size
+                self.best['ssa_path'] = ssa_path
+
+            # check if we have run out of time
+            if (self.max_time is not None) and (time.time() > t0 + self.max_time):
```

Comment:
`if self.executor is not None:` can be the exterior if here.

Reply:
Ah yes good point.